File: Sources/custom_adapter/TopOnAdapter/Objc/6.5.x/Splash/AlxTopOnSplashDelegate.m
//
//  AlxTopOnSplashDelegate.m
//  AlxAdsOCDemo
//

#import "AlxTopOnSplashDelegate.h"

static NSString *const TAG = @"AlxTopOnSplashDelegate";

@implementation AlxTopOnSplashDelegate

#pragma mark - AlxSplashAdDelegate

- (void)splashAdDidLoad:(AlxSplashAd *)ad {
    NSLog(@"%@: splashAdDidLoad", TAG);
    self.splashAd = ad;

    double price = [ad getPrice];
    NSMutableDictionary *adExtra = [NSMutableDictionary dictionary];

    if (price > 0) {
        NSString *priceStr = [NSString stringWithFormat:@"%.2f", price];
        adExtra[ATAdSendC2SBidPriceKey] = priceStr;
        adExtra[ATAdSendC2SCurrencyTypeKey] = @(ATBiddingCurrencyTypeUS);
        NSLog(@"%@: splashAdDidLoad: price = %@", TAG, priceStr);
    }

    adExtra[kATAdAssetsCustomObjectKey] = ad;
    [self notifySplashLoaded:adExtra];
}

- (void)splashAdDidFailToLoad:(AlxSplashAd *)ad didFailWithError:(NSError *)error {
    NSLog(@"%@: splashAdDidFailToLoad: %@", TAG, error.localizedDescription);
    [self notifyLoadFailed:error];
}

- (void)splashAdDidShow:(AlxSplashAd *)ad {
    NSLog(@"%@: splashAdDidShow", TAG);
    [self notifyAdShow];
}

- (void)splashAdDidClick:(AlxSplashAd *)ad {
    NSLog(@"%@: splashAdDidClick", TAG);
    [self notifyAdClick];
}

- (void)splashAdDidClose:(AlxSplashAd *)ad {
    NSLog(@"%@: splashAdDidClose", TAG);
    [self notifyAdClosed];
}

- (void)splashAdRenderDidFail:(AlxSplashAd *)ad didFailWithError:(NSError *)error {
    NSLog(@"%@: splashAdRenderDidFail: %@", TAG, error.localizedDescription);
    [self notifyAdShowFailed:error];
}

- (void)splashAdCountdown:(AlxSplashAd *)ad countdown:(NSInteger)countdown {
    NSLog(@"%@: splashAdCountdown: %ld", TAG, (long)countdown);
    if (self.adStatusBridge) {
        SEL selector = NSSelectorFromString(@"atOnSplashAdCountdownTime:");
        if ([self.adStatusBridge respondsToSelector:selector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
            [self.adStatusBridge performSelector:selector withObject:@(countdown)];
#pragma clang diagnostic pop
        }
    }
}

#pragma mark - Helper Notifications

- (void)notifySplashLoaded:(NSDictionary *)adExtra {
    if (self.adStatusBridge) {
        SEL selector = NSSelectorFromString(@"atOnSplashAdLoadedExtra:");
        if ([self.adStatusBridge respondsToSelector:selector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
            [self.adStatusBridge performSelector:selector withObject:adExtra];
#pragma clang diagnostic pop
        }
    }
}

- (void)notifyLoadFailed:(NSError *)error {
    if (self.adStatusBridge) {
        SEL selector = NSSelectorFromString(@"atOnAdLoadFailed:adExtra:");
        if ([self.adStatusBridge respondsToSelector:selector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
            [self.adStatusBridge performSelector:selector withObject:error withObject:@{}];
#pragma clang diagnostic pop
        }
    }
}

- (void)notifyAdShow {
    if (self.adStatusBridge) {
        SEL selector = NSSelectorFromString(@"atOnAdShow:");
        if ([self.adStatusBridge respondsToSelector:selector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
            [self.adStatusBridge performSelector:selector withObject:@{}];
#pragma clang diagnostic pop
        }
    }
}

- (void)notifyAdClick {
    if (self.adStatusBridge) {
        SEL selector = NSSelectorFromString(@"atOnAdClick:");
        if ([self.adStatusBridge respondsToSelector:selector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
            [self.adStatusBridge performSelector:selector withObject:@{}];
#pragma clang diagnostic pop
        }
    }
}

- (void)notifyAdClosed {
    if (self.adStatusBridge) {
        SEL selector = NSSelectorFromString(@"atOnAdClosed:");
        if ([self.adStatusBridge respondsToSelector:selector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
            [self.adStatusBridge performSelector:selector withObject:@{}];
#pragma clang diagnostic pop
        }
    }
}

- (void)notifyAdShowFailed:(NSError *)error {
    if (self.adStatusBridge) {
        SEL selector = NSSelectorFromString(@"atOnAdShowFailed:extra:");
        if ([self.adStatusBridge respondsToSelector:selector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
            [self.adStatusBridge performSelector:selector withObject:error withObject:@{}];
#pragma clang diagnostic pop
        }
    }
}

@end

File: Sources/custom_adapter/TopOnAdapter/Swift/6.5.x/Splash/AlxTopOnSplashDelegate.swift
//
//  AlxTopOnSplashDelegate.swift
//  AlxAdsDemo
//

import Foundation
import UIKit
import AnyThinkSDK
import AlxAds

@objc(AlxTopOnSplashDelegate)
public class AlxTopOnSplashDelegate: NSObject, AlxSplashAdDelegate {
    
    private static let TAG = "AlxTopOnSplashDelegate"
    
    @objc public var adStatusBridge: ATSplashAdStatusBridge?
    @objc public weak var splashAd: AlxSplashAd?
    
    // MARK: - AlxSplashAdDelegate
    
    public func splashAdDidLoad(_ ad: AlxSplashAd) {
        NSLog("%@: splashAdDidLoad", AlxTopOnSplashDelegate.TAG)
        self.splashAd = ad
        
        let price = ad.getPrice()
        var adExtra: [AnyHashable: Any] = [:]
        
        if price > 0 {
            let priceStr = String(format: "%.2f", price)
            adExtra[ATAdSendC2SBidPriceKey] = priceStr
            adExtra[ATAdSendC2SCurrencyTypeKey] = NSNumber(value: 1)
            NSLog("%@: splashAdDidLoad: price = %@", AlxTopOnSplashDelegate.TAG, priceStr)
        }
        
        adExtra[kATAdAssetsCustomObjectKey] = ad
        self.notifySplashLoaded(adExtra: adExtra)
    }
    
    public func splashAdDidFailToLoad(_ ad: AlxSplashAd, didFailWithError error: Error) {
        NSLog("%@: splashAdDidFailToLoad: %@", AlxTopOnSplashDelegate.TAG, error.localizedDescription)
        self.notifyLoadFailed(error: error)
    }
    
    public func splashAdDidShow(_ ad: AlxSplashAd) {
        NSLog("%@: splashAdDidShow", AlxTopOnSplashDelegate.TAG)
        self.notifyAdShow()
    }
    
    public func splashAdDidClick(_ ad: AlxSplashAd) {
        NSLog("%@: splashAdDidClick", AlxTopOnSplashDelegate.TAG)
        self.notifyAdClick()
    }
    
    public func splashAdDidClose(_ ad: AlxSplashAd) {
        NSLog("%@: splashAdDidClose", AlxTopOnSplashDelegate.TAG)
        self.notifyAdClosed()
    }
    
    public func splashAdRenderDidFail(_ ad: AlxSplashAd, didFailWithError error: Error) {
        NSLog("%@: splashAdRenderDidFail: %@", AlxTopOnSplashDelegate.TAG, error.localizedDescription)
        self.notifyAdShowFailed(error: error)
    }
    
    public func splashAdCountdown(_ ad: AlxSplashAd, countdown: Int) {
        NSLog("%@: splashAdCountdown: %d", AlxTopOnSplashDelegate.TAG, countdown)
        if let bridge = self.adStatusBridge {
            let selector = NSSelectorFromString("atOnSplashAdCountdownTime:")
            if bridge.responds(to: selector) {
                bridge.perform(selector, with: countdown as NSNumber)
            }
        }
    }
    
    // MARK: - Dynamic Invocation Helper Methods
    
    private func notifySplashLoaded(adExtra: [AnyHashable: Any]) {
        if let bridge = self.adStatusBridge {
            let selector = NSSelectorFromString("atOnSplashAdLoadedExtra:")
            if bridge.responds(to: selector) {
                bridge.perform(selector, with: adExtra)
            }
        }
    }
    
    private func notifyLoadFailed(error: Error) {
        if let bridge = self.adStatusBridge {
            let selector = NSSelectorFromString("atOnAdLoadFailed:adExtra:")
            if bridge.responds(to: selector) {
                bridge.perform(selector, with: error, with: nil)
            }
        }
    }
    
    private func notifyAdShow() {
        if let bridge = self.adStatusBridge {
            let selector = NSSelectorFromString("atOnAdShow:")
            if bridge.responds(to: selector) {
                let emptyDict: [AnyHashable: Any] = [:]
                bridge.perform(selector, with: emptyDict)
            }
        }
    }
    
    private func notifyAdClick() {
        if let bridge = self.adStatusBridge {
            let selector = NSSelectorFromString("atOnAdClick:")
            if bridge.responds(to: selector) {
                let emptyDict: [AnyHashable: Any] = [:]
                bridge.perform(selector, with: emptyDict)
            }
        }
    }
    
    private func notifyAdClosed() {
        if let bridge = self.adStatusBridge {
            let selector = NSSelectorFromString("atOnAdClosed:")
            if bridge.responds(to: selector) {
                let emptyDict: [AnyHashable: Any] = [:]
                bridge.perform(selector, with: emptyDict)
            }
        }
    }
    
    private func notifyAdShowFailed(error: Error) {
        if let bridge = self.adStatusBridge {
            let selector = NSSelectorFromString("atOnAdShowFailed:extra:")
            if bridge.responds(to: selector) {
                bridge.perform(selector, with: error, with: nil)
            }
        }
    }
}
